ccf_transaction_processor/transaction_processor/pdo_tp.cpp

```diff
@@ -0,0 +1,624 @@
+            auto contract_view = tx.get_view(contracttable);
+            auto enclave_view = tx.get_view(enclavetable);
+
+            // ensure that contract was previously registered
+            auto contract_r = contract_view->get(in.contract_id);
+            if (!contract_r.has_value()) {
+                return make_error(
+                    jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Contract not yet registered");
+            }
+            auto contract_info = contract_r.value();
+
+            //ensure the contract owner is the one adding enclaves to contract
+            if (in.contract_creator_verifying_key_PEM != contract_info.contract_creator_verifying_key_PEM){
+                return make_error(
+                    jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Enclave can only be added by the contract owner");
+            }
+
+            // Verify Pdo transaction signature
+            if (!verify_pdo_transaction_signature_add_enclave(in.signature, in.contract_creator_verifying_key_PEM, \
+                    in.contract_id, in.enclave_info)){
+                return make_error(
+                    jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Invalid PDO payload signature");
+            }
+
+            // Parse the enclave info json string
+            std::vector<ContractEnclaveInfo> enclave_info_array;
+            try {
+                auto j = nlohmann::json::parse(in.enclave_info);
+                enclave_info_array = j.get<std::vector<ContractEnclaveInfo>>();
+            }
+            catch(...){
+                return make_error(
+                    jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Unable to parse ContractEnclaveInfo Json");
+            }
+
+            // verify enclave_info_array
+
+            //unsure if this check is needed, but keeping it for now
+            if (enclave_info_array.size() == 0) {
+                return make_error(
+                    jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Provide at least one encalve to add to contract");
+            }
+
+            // check each element of the array
+            for (auto enclave_info_temp: enclave_info_array){
+
+                // check contract id contained in enclave info
+                if (enclave_info_temp.contract_id != contract_info.contract_id){
+                    return make_error(
+                        jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Enclave info has invalid contract id");
+                }
+
+                //ensure enclave is registered
+                auto enclave_r = enclave_view->get(enclave_info_temp.contract_enclave_id);
+                if (!enclave_r.has_value()){
+                    return make_error(
+                        jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Enclave not yet registered");
+                }
+                auto enclave_info_ledger = enclave_r.value();
+
+                //check if this enclave has already been added to the contract
```

Comment:
thank you for the question. I would prefer throwing an error if the contract creator tries to add the same enclave twice. A well-behaved contract creator is not expected to do this anyway. 

Also, this is the same implementation in Sawtooth based PDO TP. (see contract_registry_handler.py)

Reply:
adding an enclave involves several checks, and the add operation fails even if any one of these checks fails. I don't see how the order of performing these mutually independent checks speeds up anything on average if the error is one that occurs at random

Reply:
I don't fully follow. How is the error occurring at random? 

Sure, there are several checks that need to pass in order to add an enclave. But in the scenario that a given enclave _has already been added_ and we want to return an error, the only check that truly matters is this one: https://github.com/hyperledger-labs/private-data-objects/blob/4cac5c5547b9a719ef1d436df314e95a58c7d44d/ccf_transaction_processor/transaction_processor/pdo_tp.cpp#L258

The other two checks will pass in this scenario, but they don't matter since we end up returning an error anyway in the end, so we have performed redundant checks. I'm not going to push that this be changed in this PR simply because performance isn't a priority at this moment. But this seems like a fairly straight-forward change that could potentially save us some operations.

Reply:
The scenario "that a given enclave has already been added" is one of the possible scenarios for which invalidity is checked. Scenarios such as "contract not registered" or "Enclave not registered" all can cause this add_enclave_transaction to fail. 

This was my motivation to say that I don't see why one check must be prioritized over the other. All error scenarios are best treated equally likely to happen.